### PR TITLE
Modified init and removed all instances of add_or_remove_edge method to work with new format

### PR DESCRIPTION
**Pull request by aditrishukla7**
Modified FPT2Block init and removed all instances of add_or_remove_edge method to work with new format. Files edited were:

- tests/test_label_smoothing_init.py
- src/layer2/modeling/modeling_fpt2.py
- src/layer2/prune/edge_pruning_mezo/fpt2_gp.py

### CODE REVIEW

File: src/layer2/modeling/modeling_fpt2.py [hunk 1]
```diff
@@ -525,235 +550,7 @@ def __init__(
         
         mlp_write_common_mask = torch.zeros((self.n_writers, 1), dtype=self._dtype)
         mlp_write_common_mask[self.mlp_writer_offset, 0] = 1
-        self.register_buffer("mlp_write_common_mask", mlp_write_common_mask)
```
Comment (HELLO360360360) [inline src/layer2/modeling/modeling_fpt2.py:527]: Hi Aditri! It looks fine to me. Great work! I have a question, however. Why does the code underneath have to be deleted? This code kind of seems important and would like some clarification. Thanks!

File: src/layer2/modeling/modeling_fpt2.py [hunk 2]
```diff
@@ -477,43 +479,67 @@ def __init__(
...
+        if initial_scores is None:
+            self.q_read_log_alphas.data.normal_(mean=10.0, std=0.01)
+            self.k_read_log_alphas.data.normal_(mean=10.0, std=0.01)
+            self.v_read_log_alphas.data.normal_(mean=10.0, std=0.01)
+            self.mlp_read_log_alphas.data.normal_(mean=10.0, std=0.01)
+        else:
+            def score_to_log_alpha(score):
+                return 10.0 + (score - 0.5) * 5.0
+
+            self.q_read_log_alphas.data.fill_(-10.0)
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py:507]: Is there a reason behind using -10 as the default value for nodes that could not be found?

File: src/layer2/modeling/modeling_fpt2.py (hunk 1 above)
Comment (MustafaKhan0) [reply src/layer2/modeling/modeling_fpt2.py:527]: Completely agree - why is the attn_read, write, mlp_read, write, and literal forward method deleted?

File: src/layer2/modeling/modeling_fpt2.py [hunk 3]
```diff
@@ -816,75 +614,45 @@ class FPT2ModelOutput(ModelOutput):
...
-            sparsity_lambda_nodes_1 = torch.tensor([0.0], dtype=self._dtype)
-            self.register_buffer("sparsity_lambda_edges_1", sparsity_lambda_edges_1)
-            self.register_buffer("sparsity_lambda_nodes_1", sparsity_lambda_nodes_1)
-        else:
-            self.sparsity_lambda_edges_1 = nn.Parameter(torch.tensor([0.0], dtype=self._dtype))
-            self.sparsity_lambda_nodes_1 = nn.Parameter(torch.tensor([0.0], dtype=self._dtype))
-        self.sparsity_lambda_edges_2 = nn.Parameter(torch.tensor([0.0], dtype=self._dtype))
-        self.sparsity_lambda_nodes_2 = nn.Parameter(torch.tensor([0.0], dtype=self._dtype))
-
-        # Initialize weights and apply final processing
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py:887]: Why are there so many changes to the init method? I'm really unsure as to why we're deleting lambda parameters, and a bunch of other config values?

File: src/layer2/modeling/modeling_fpt2.py [hunk 4]
```diff
@@ -1064,29 +832,6 @@ def get_edges(self):
...
-        elif to_node.startswith("m"):
-            layer_idx = int(to_node[1:])
-            old_value = self.h[layer_idx].set_mlp_mask_value(from_idx, value)
-        else:
-            parts = to_node.split(".")
-            layer_idx = int(parts[0][1:])
-            head_idx = int(parts[1][1:])
-            qkv = parts[2]
-            old_value = self.h[layer_idx].set_attn_mask_value(from_idx, head_idx, qkv, value)
-        return old_value
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py]: We should probably leave this method here (since I believe it is from the original codebase?) even if we aren't using it.

File: src/layer2/modeling/modeling_fpt2.py [hunk 5]
```diff
@@ -477,43 +479,67 @@ def __init__(
...
+                        self.mlp_read_log_alphas.data[from_idx] = score_to_log_alpha(score)
+
+                    for head_idx in range(self.n_head):
+                        attn_reader_as_writer_name = f"a{layer_idx}.h{head_idx}"
+                        if attn_reader_as_writer_name in to_scores:
+                            score = to_scores[attn_reader_as_writer_name]
+                            self.q_read_log_alphas.data[from_idx, head_idx] = score_to_log_alpha(score)
+                            self.k_read_log_alphas.data[from_idx, head_idx] = score_to_log_alpha(score)
+                            self.v_read_log_alphas.data[from_idx, head_idx] = score_to_log_alpha(score)
+
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py:529]: I think a more detailed documentation on the logic behind this code would be really helpful. Additionally, some documentation on the schema for the inputs here would be helpful.

File: src/layer2/modeling/modeling_fpt2.py [hunk 6]
```diff
@@ -477,43 +479,67 @@ def __init__(
...
+            self.q_read_log_alphas.data.fill_(-10.0)
+            self.k_read_log_alphas.data.fill_(-10.0)
+            self.v_read_log_alphas.data.fill_(-10.0)
+            self.mlp_read_log_alphas.data.fill_(-10.0)
+
+            for from_idx in range(self.n_writers):
+                from_node = writer_idx_to_name(from_idx, config.n_layer, self.n_head, with_embedding_nodes)
+                if from_node in initial_scores:
+                    to_scores = initial_scores[from_node]
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py:515]: I'm kinda confused about this code:
- If I understand the initial_scores / edge_scores value correctly, then the dictionary follows the format of ```{'node_whose_mask_is_being_defined' : {'other_node_in_computational_graph' : importance_score}}```
- By iterating through the number of writers, and using the writers, you're iterating through the other nodes in the computational graph that write to the residual stream. 
- Therefore, in the __init__ method for a particular transformer block, we wouldn't be iterating through the entire initial_scores dictionary - rather just looking at a few particular ones, whose masks are stored in that transformer block (i.e. that layer's MLP and attention heads)

File: src/layer2/modeling/modeling_fpt2.py [hunk 7]
```diff
@@ -477,43 +479,67 @@ def __init__(
...
+
+        # --- MASKS FOR INFORMATION FLOW (Unchanged) ---
         attn_read_common_mask = torch.zeros(self.n_writers, dtype=self._dtype)
         attn_read_common_mask[:self.attn_writer_offset] = 1
         attn_read_common_mask = attn_read_common_mask.unsqueeze(1)
         self.register_buffer("attn_read_common_mask", attn_read_common_mask)
         
         attn_write_common_mask = F.pad(
-            torch.eye(self.n_head, dtype=torch.float32).to(self._dtype), # eye does not support bfloat16
+            torch.eye(self.n_head, dtype=torch.float32).to(self._dtype),
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py]: We should probably leave that comment there

File: src/layer2/modeling/modeling_fpt2.py [hunk 8]
```diff
@@ -1454,24 +1199,20 @@ class FPT2LMHeadModel(FPT2PreTrainedModel):
...
             disable_linear_regularization_term=disable_linear_regularization_term,
+            initial_scores=initial_scores,
         )
         self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
-
-        # Model parallel
-        self.model_parallel = False
-        self.device_map = None
-
-        # Initialize weights and apply final processing
```
Comment (MustafaKhan0) [inline src/layer2/modeling/modeling_fpt2.py]: Is there a reason to remove this code?

File: src/layer2/prune/edge_pruning_mezo/fpt2_gp.py
Comment (MustafaKhan0) [inline src/layer2/prune/edge_pruning_mezo/fpt2_gp.py:1]: Would be really helpful if we could make the same changes to the files for the other tasks (gt, ioi, etc).

**Review by MustafaKhan0 — CHANGES_REQUESTED** [review]
I think there's a decent amount of confusing changes, which could use more documentation in the form of comments or in the PR description. 

Also, have you run the any of the main experiments with these changes? If so, what did it generally look like?

File: src/layer2/prune/edge_pruning_mezo/fpt2_gp.py [hunk 9]
```diff
@@ -566,11 +577,19 @@ def main():
 
     raw_datasets = load_datasets(data_args.dataset_path, data_args.max_train_samples, data_args.max_eval_samples, train_split=data_args.train_split)
     n_train = len(raw_datasets["train"])
+
+    initial_scores = None
+    if data_args.label_smoothing_alpha is not None and data_args.initial_mask_path is not None:
+        print("Applying label smoothing initialization...")
+        initial_mask = load_mask(data_args.initial_mask_path)
+        z_star = compute_z_star(initial_mask, alpha=data_args.label_smoothing_alpha)
```
Comment (aarushM923) [inline src/layer2/prune/edge_pruning_mezo/fpt2_gp.py]: This line is calling compute_z_star from label_smoothing_utils file in layer1. That means Layer 2 is applying label smoothing internally rather than passing a dictionary of smoothed scores in, since compute_z_star expects a binary mask field and then expands it into smoothed scores.

File: src/layer2/prune/edge_pruning_mezo/fpt2_gp.py [hunk 10]
```diff
@@ -69,6 +69,9 @@
     )
 )   # Very hacky but the imports are annoying otherwise
 from modeling_fpt2 import FPT2LMHeadModel
+sys.path.append(os.path.join(os.getcwd(), "src/layer1/soft_edge_mask"))
+from label_smoothing_utils import load_mask, compute_z_star, compute_edge_scores
```
Comment (aarushM923) [inline src/layer2/prune/edge_pruning_mezo/fpt2_gp.py]: Since Layer 2 isn't supposed to be smoothing, this import is partially unused, so it should be deleted.

Only load_mask and compute_edge_scores are still needed.

**Review by aarushM923 — CHANGES_REQUESTED** [review]
I left two comments related to the use of compute_z_star. Based on my understanding, label smoothing and compute_z_star should stay in Layer 1, while Layer 2 should take in the smoothed edge scores. When I was coding the root main.py file from your branch, I was running into logic errors due to this.
